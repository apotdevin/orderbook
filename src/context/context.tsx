import { FC, createContext, useContext, useReducer } from 'react';

const getStep = (x: number) => {
  switch (x) {
    case 1:
      return 0.5;
    case 2:
      return 1;
    case 3:
      return 2.5;
    case 4:
      return 5;
    case 5:
      return 10;
    case 6:
      return 25;
    case 7:
      return 50;
    case 8:
      return 100;
    case 9:
      return 250;
    case 10:
      return 500;
    case 11:
      return 1000;
    case 12:
      return 2500;
    default:
      return 0.5;
  }
};

type State = {
  group: number;
  groupStep: number;
  limit: number;
  spread: number;
};

type ActionType =
  | {
      type: 'change';
      spread: number;
    }
  | { type: 'changeLimit'; limit: number }
  | { type: 'changeGroup'; group: number };

type Dispatch = (action: ActionType) => void;

export const StateContext = createContext<State | undefined>(undefined);
export const DispatchContext = createContext<Dispatch | undefined>(undefined);

const stateReducer = (state: State, action: ActionType): State => {
  switch (action.type) {
    case 'change':
      return { ...state, spread: action.spread };
    case 'changeLimit':
      return { ...state, limit: Math.min(Math.max(action.limit, 10), 100) };
    case 'changeGroup': {
      const group = Math.min(Math.max(action.group, 1), 12);
      return {
        ...state,
        group,
        groupStep: getStep(group),
      };
    }
    default:
      return state;
  }
};

const ContextProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(stateReducer, {
    spread: 0,
    group: 1,
    groupStep: getStep(1),
    limit: 15,
  });

  return (
    <DispatchContext.Provider value={dispatch}>
      <StateContext.Provider value={state}>{children}</StateContext.Provider>
    </DispatchContext.Provider>
  );
};

const useContextState = () => {
  const context = useContext(StateContext);
  if (context === undefined) {
    throw new Error('useContextState must be used within a ContextProvider');
  }
  return context;
};

const useContextDispatch = () => {
  const context = useContext(DispatchContext);
  if (context === undefined) {
    throw new Error('useContextDispatch must be used within a ContextProvider');
  }
  return context;
};

export { ContextProvider, useContextState, useContextDispatch };
