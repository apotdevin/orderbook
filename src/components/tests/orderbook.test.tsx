import { render } from '@testing-library/react';
import { ContextProvider } from 'src/context/context';
import WS from 'jest-websocket-mock';
import OrderBook from '../orderbook';

new WS('ws://localhost:3000');

describe('Components', () => {
  describe('Orderbook', () => {
    test('matches snapshot', () => {
      const { asFragment } = render(
        <ContextProvider>
          <OrderBook url={'ws://localhost:3000'} />
        </ContextProvider>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
