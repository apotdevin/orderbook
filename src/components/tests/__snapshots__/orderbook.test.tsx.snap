// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Orderbook matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="orderbook__wrapper-ypat8y-1 gKWLzb"
  >
    Trying to connect to websocket...
    <div
      class="loading__Wrapper-e7lqu1-0 dNIoQG"
    >
      <span
        class="css-0"
      >
        <span
          class="css-t5z7rz"
        />
        <span
          class="css-180eqmy"
        />
        <span
          class="css-10qunde"
        />
        <span
          class="css-1ct7tzo"
        />
        <span
          class="css-1cubgjj"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
